Add FluidIngredient

3TUSK opened the pull request:
Title.
In short, an `Ingredient` that matches `ItemStack`s based on contents in their `IFluidHandlerItem`.

This also serves as an alternative approach of #3332.

I encountered its use case while I was writing [this mod](https://github.com/3TUSK/MooncakeCraft). Detail about the use case is simple: in real life, mooncake's crust is made from dough, and thus I have to implement such a process via some kind of `IRecipe`. Then I realize that I have to write a recipe that accepts arbitrary `ItemStack` that contains water, i.e. has `IFluidHandlerItem` capability and contains specific amount of water.  
For now I am using water bottle (`minecraft:potion` with `IngredientNBT`), but I received some feedback saying that it is not realistic as the glass bottle is consumed. 

Other use cases may exist, but for now I can only recall Railcraft (wooden tie, requires creosote, i.e. #3332) and Immersive Engineering (treated wood, also requires creosote).

Some details:
* A test mod that boils down my real use case is included. You can find example usage in its `_constants.json`.
* There are ~~three~~ ~~two~~ three logics available for choosing: ~~`EXACT`, `EXCEED` and `VOIDING`~~ ~~`EXACT`and `EXCEED`~~ `EXACT`, `EXCEEDED`, and `NON_VOIDING`.
  * `EXACT` means that `FluidStack::isFluidStackIdentical == true`, i.e. amount sensitive.
  * `EXCEED`, the default behavior when not specified, only requires `FluidStack::isFluidStackEqual == true` and the amount of fluid supplied is *exceeded*, or at least equal.
  * ~~`VOIDING`, same as `EXCEED`, except it will try voiding all fluid remained in container. Think about a recipe that asks for 500 mB water and user gives a vanilla water bucket: user will get empty bucket.~~ Removed as no actual use case is present.
  * `NON_VOIDING` is a strict version of `EXCEEDED`, as it will enables the check of "if this container can actually be drained with specific amount of fluid". Example: when a recipe asks for 500 mB water, and user gives a vanilla water bucket, `EXCEEDED` will waste the 500 mB water and give user the crafting result; `NON_VOIDING` will fail on matching the recipe.
* Remaining container is handled by `IFluidHandlerItem::getContainer`.
* I left `::getMatchingStacks` and `::getValidItemStacksPacked` as stub (using `ItemStack[0]` and `IntLists.EMPTY`). However, this breaks vanilla recipe book when it tries displaying the phantom recipe. I have considered replacing it with `ItemStack[1]` that contains a forge universal bucket, but not sure whether it is appropriate or not. Inputs are welcome.

Appendix I: Since it's `Ingredient`, modder can also use it in `minecraft:crafting_shaped` and `minecraft:crafting_shapeless`. Will we allow that happen?

~~Appendix II: see #4359 for issue in `::getMatchingStacks`.~~ Fixed

~~Appendix III: see #4472 for issue within shapeless recipe matching and/or `getRemainingItems` algorithm.~~ ~~Addressed; only actual implementation is pending.~~ Fixed

Appendix IV: see #4516 and #4519 for underlying issue within `::getMatchingStacks`, `::getValidItemStacksPacked` and other affect methods, which are currently blocking this feature from being properly implemented.

Appendix V: a new interface, `ICustomContainerCallback` is added to provide an abstraction layer on how `FluidIngredient` handles getting container item. This may be used in other custom `Ingredient`.

------

First time of sending pull request to forge, not sure what I missed, feedbacks are welcome.

3TUSK committed:
Add FluidIngredient, supported by ShapedOreRecipe, and a test mod for it

Also serves as Ingredient version of MinecraftForge/MinecraftForge#3332

3TUSK committed:
Also covers FluidIngredient support in ShapedlessOreRecipe

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=4449) <br/>All committers have signed the CLA.

mezz commented:
Looks good!
One thing, I do not really see the point of having a `VOIDING` option. I think that behavior should be left up to the container, like the bucket would have to void its excess anyway.

KnightMiner commented:
Would it be worth adding a default recipe using the milk fluid to make cake for Forge? I know forge does not add milk by default, but if a new condition was added like the item one (something like `forge:fluid_exists`), it can be made to only load when milk is detected, saving on duplicate modded container milk recipes.

3TUSK commented:
@mezz 
Removing `VOIDING` option can simplify the enum into a boolean field, which would be nice - if no actual use case is present. Awaiting more comments.
Also - can this supersede #3332?

@KnightMiner 
Sounds interesting... personally I'd say it's worth. Still, awaiting more comments.

mezz commented:
I think you should keep it as an 2-value enum, boolean parameters can be confusing.

3TUSK commented:
I recall another potential issue: 
By using `Ingredient`, one can also use this `FluidIngredient` in `minecraft:crafting_shaped` and `minecraft:crafting_shapeless` (based on my understanding). I am not sure if there is a way to limits the usage of `FluidIngredient` to only `forge:ore_shaped` and `forge:ore_shapeless`, but without patching the vanilla recipe classes, they cannot handle `FluidIngredient` correctly.  
Correct me if there is misinterpretation.

liach commented:
you might need more debug to make sure the shapeless recipe is absolutely correct.

3TUSK committed:
Remove VOIDING strategy; fix shapeless recipe handling; extra edge case test; tooltip for test items

3TUSK commented:
@mezz `VOIDING` is removed, after some private discussion.

@liach An edge case shapeless recipe involving 4 * 1000 mB water and 4 * 1000 mB lava is included. Works as intended.

liach commented:
2*1000+750

liach commented:
When you need one bucket of watet and 3/4 bucket of water, your result should be 2 empty buckets or two 1/4 buckets for your incorrect algorithm on remaining item processing.

3TUSK commented:
2 * 1000 mB + 1 * 750 mB of which fluid(s).

3TUSK commented:
> one bucket of watet and 3/4 bucket of water

If you give 2 vanilla water buckets then 2 empty buckets in return is intended behavior.

> two 1/4 buckets

I am not sure what you mean by this.

liach commented:
@3TUSK 
1. First, you need to add a type of fluid container that is specific to each millibucket (let's say, 1000 millibuckets), so the bug can be presented.
2. This bug only exists in shapeless recipes.
3. Add a recipe that needs 1 bucket of water and another 750 millibuckets of water (actually, any fluid is applicable).
4. Correct result is that one specific container is empty (or has 1000 millibuckets drained) and the other one has 250 millibuckets left (or has 750 millibuckets drained).
5. However, your result would be 2 empty buckets (or 2 * 1000 millibuckets drained) or two 250 millibuckets left (or has 2 * 750 millibuckets drained) for your incorrect algorithm on remaining item processing in the shapeless recipes.

3TUSK commented:
> ~~your result would be 2 empty buckets (or 2 * 1000 millibuckets drained)~~

~~If you use vanilla water bucket then it is intended behavior.~~  


> ~~two 250 millibuckets left (or has 2 * 750 millibuckets drained)~~

~~I don't believe that's possible.~~

Edit: see my comment below.

liach commented:
Just add that test and prove it works.

3TUSK commented:
Do you really think you can drain 250mB water out of an vanilla bucket?

No, I still don't think you have a valid argument; but further investigation on `FluidBucketWrapper` reveals that I made a wrong assumption on somewhere else though...

Appendix:
For anyone wonder, my implementation has issue that it doesn't check if it really drains fluid from container. This issue will occur on vanilla bucket, forge universal bucket, forestry tin can, or anything that can only holds, fill and drain a fixed amount of fluid.  
Solution... incoming.

liach commented:
> First, you need to add a type of fluid container that is specific to each millibucket (let's say, 1000 millibuckets), so the bug can be presented.

Read this. Almost all mods do that. Just add one of that type of item and add the recipe to test.

liach commented:
Actually nevermind, this one is fine.

liach commented on src/main/java/net/minecraftforge/fluids/FluidIngredient.java:
```diff
@@ -0,0 +1,145 @@
+    @Override
+    public ItemStack[] getMatchingStacks()
+    {
+        return new ItemStack[0];
```
We might need some improvement on this one, I guess...

3TUSK replied:
>  I have considered replacing it with ItemStack[1] that contains a forge universal bucket, but not sure whether it is appropriate or not. Inputs are welcome.

3TUSK replied:
Actually no, the universal bucket is not enabled by default...

3TUSK replied:
@mezz Is there any chance to add a new Item for displaying `FluidStack` here, or I have to patch vanilla recipe book to allow `ItemStack[0]` and `IntLists.EMPTY`?

Edit: See #4359.

mezz replied:
It would be more user friendly to see the items that hold the fluid here.
I am not sure what the most efficient way to find those items is in Forge, but I think it's worth looking into.

loordgek replied:
is it a a option to bring fluidBucketRegistry back ??

3TUSK replied:
You mean the `FluidContainerRegistry`? I am afraid that it's impossible.

mezz replied:
We could listen to `AttachCapabilitiesEvent<ItemStack>` to check every itemstack with capabilities and create a list of everything that has the fluid item capability. Besides that I don't really have a good idea of how to do it.

3TUSK replied:
Does that guarantee that we can get at least 1 valid container?

mezz replied:
If a fluid recipe is registered with no valid container to work with it, that is an issue. You could hide it from the recipe book in that case (using `IRecipe#isDynamic`?) since it is an uncraftable recipe.

3TUSK replied:
Will see if that works. Second question: doesn't maintaining a collection of known containers consume memory?

mezz replied:
It does, but not a lot. Mods will be able to make use of it too, a few JEI integrations for fluid-handling machines already search through all the items for ones with fluid capabilities.

Alpvax replied:
How would that work with items that dynamically expose the fluid capability? Such as an item which can hold liquid if there is a certain upgrade applied? Chances are that the cap won't be exposed during AttachCapabilitiesEvent.
Having an additional way to add itemstacks to the list could be useful (could add an itemstack with the upgrade applied).

mezz replied:
For simplicity I don't think this should support dynamically exposed fluid capabilities. It's just meant to give the player a general idea of what can be used in the recipe.

3TUSK replied:
I have an idea: specify that as a part of `FluidIngredient`? Like this:

````json
{
  "type": "forge:fluid",
  "fluid": "minecraft:water",
  "amount": 1000,
  "matching": "exact",
  "container": "forge:bucket"
}

mezz replied:
I would prefer the more automatic way, it can show most valid items without any effort.

Alpvax replied:
@3TUSK Adding it to the FluidIngredient would mean that some people might think that it had to be a forge:bucket in said recipe, rather than any container

liach replied:
I think this is great now.

3TUSK committed:
Add NON_VOIDING strategy (see commit message); rename EXCEED to EXCEEDED

The logic behind NON_VOIDING strategy is as follow: for an vanilla water
bucket, if a recipe asks for 500 mB water, the bucket cannot provide
(i.e. `drain(500, true)` returns null), even though it does have 1000 mB
water. Thus, in order to drain 500 mB water from vanilla bucket, waste
cannot be avoided. NON_VOIDING enables the strict check so that a
`FluidIngredient` only matches those containers that can really be drained
with specified amount of fluid.

LatvianModder commented:
You can reduce this
```java
Preconditions.checkNotNull(fluidStack, "Null FluidStack is not permitted in FluidIngredient");
Preconditions.checkNotNull(strategy, "MatchingStrategy cannot be null");
this.fluidStack = fluidStack;
this.strategy = strategy;
```
to just 
```java
this.fluidStack = Preconditions.checkNotNull(fluidStack, "Null FluidStack is not permitted in FluidIngredient");
this.strategy = Preconditions.checkNotNull(strategy, "MatchingStrategy cannot be null");
```

3TUSK committed:
Take the return value, simplify code a bit

3TUSK commented:
@LatvianModder Adapted.

LexManos commented:
Gunna tag @KingLemming on this one as Fluids are his an cpw's baby. They are in control of any design decisions.

3TUSK commented:
Update: 
There is an issue regarding `ShapelessRecipes::match`, which may affect all current implementation of shapeless recipe in minecraft & minecraftforge. See #4472 for details.  
The issue also reveals that the current implementation in this pull request is still wrong.

KnightMiner commented on src/main/java/net/minecraftforge/oredict/ShapedOreRecipe.java:
```diff
@@ -143,6 +147,41 @@ protected boolean checkMatch(InventoryCrafting inv, int startX, int startY, bool
+            if (!itemStack.isEmpty())
+            {
+                Ingredient ingredient = input.get(index);
+                if (ingredient instanceof FluidIngredient)
```
Could we make this an interface that fluid ingredients implement and use a callback for this logic? I could see this being useful for an RF battery ingredient to have a callback to remove energy, or other similar cases where the ingredient can define special logic on crafting with it

3TUSK replied:
Could that be a part of `Ingredient` rather than an interface?

KnightMiner replied:
Well, we would need something to check for if we use the default logic of `getContainerItem()`, though I guess that logic could just be placed in ingredient itself.

3TUSK replied:
After some further thinking, I think it's better to be a separate pull request - for sake of making things clear.

liach replied:
@KnightMiner I'd suggest adding a method in ingredient class to handle remaining items instead.

KnightMiner replied:
I'd personally would just rename this PR and include it as I think the better customizability would be nice, but if you would rather do it separately that's fine (or I might try doing it later)

3TUSK replied:
@KnightMiner I think I changed my mind. `Ingredient` should be a black-box IMHO. Will kickstart on such an interface...

LexManos replied:
Problem with doing that is that it requires re-working how the remaining items are done and re-doing the matching system to align the ingredient to the input.
This is part of the reason i've written my matcher code to return the stuff that it does. But From my discussions with various people my idea is dumb and not the way they want it to go.. so I haven't implemented it.

3TUSK committed:
Disable FluidIngredient test functionality by default

liach commented:
@3TUSK However, these changes are not compatible with #4472; when that fix is merged, we need a total rewrite of Item stack recipe matcher so that it would take factors other than item type and meta into account.

3TUSK commented:
https://github.com/MinecraftForge/MinecraftForge/pull/4472#issuecomment-338363274

liach commented:
> As the original founder (per @ustc-zzzz said), I think we should fix this for `getRemainingItems` too?

Unfortunately, you did not understand my comment.

This is not an issue with remaining items. Once #4472 is merged, a fluid ingredient can never be matched in a shapeless recipe, since the `net.minecraft.client.util.RecipeItemHelper` (or `net.minecraft.world.entity.player.StackedContents` in Mojangsta's words) checks `Ingredient#getValidItemStacksPacked` to make sure if an item stack can be accepted in a 
recipe. The item stacks to compare are truncated into integers, disregarding NBT data.

Right now, your [`Ingredient#getValidItemStacksPacked`](https://github.com/MinecraftForge/MinecraftForge/pull/4449/files#diff-ebf5b9e1d3bd7eec227ec2d72fd40415R120) returns an empty `IntList`. This **always** causes the recipe matcher to assume that no item stacks can satisfy the ingredient's requirement, thus the recipes are broken.

Since the current matcher does not respect NBT, we need a whole new NBT matcher. The old one from Minecraft codebase is a mess, and is hardly understandable. (They may be using some bipartite algorithms already, based on the depth-first search in the code) Now, the problem falls onto the new recipe matcher which would respect NBT data. We will make one that takes item stacks directly instead of packed integers.

liach commented:
Update: Now #4472 is merged in c8efe29, this issue becomes disturbing. https://github.com/MinecraftForge/MinecraftForge/pull/4472#issuecomment-338371886 is my original propose for using a graph algorithm. Hope that helps.

3TUSK commented:
I would rather say it is you that did not explain what you exactly mean at first...

> The item stacks to compare are truncated into integers, disregarding NBT data.

Why use vanilla stuff then... ?

> Right now, your Ingredient#getValidItemStacksPacked returns an empty IntList

Thus https://github.com/MinecraftForge/MinecraftForge/pull/4449#discussion_r143342539. `::getValidItemStacksPacked` wasn't used that way iirc.

liach commented:
> the `net.minecraft.client.util.RecipeItemHelper` (or `net.minecraft.world.entity.player.StackedContents` in Mojangsta's words) checks `Ingredient#getValidItemStacksPacked`

This is why your fix would be invalid with the new pull by ustc-zzzz.

You can write another recipe item helper, but it would be a lot more extra work. 

> Why use vanilla stuff then... ?

I support your attempt to write one. I believe a modded recipe item helper is the only way to reach over vanilla limitations.

liach commented:
Tip: Read the code of `net.minecraft.client.util.RecipeItemHelper` to see how it works.

3TUSK commented:
> This is why your fix would be invalid with the new pull by ustc-zzzz.

In fact, I stopped updating this pull request in favor of waiting him for the solution. After all, I noticed the broken matching algorithm during testing my pull request...

LexManos commented:
Implemented a 'isSimple' system: https://github.com/MinecraftForge/MinecraftForge/commit/89db87dbfce4e4bd130f5daccac7cac495002275#diff-08bcae9db54a98c6aaa4850c80cdc3d5R110 I may expand that later to auto-hide 'non-simple' recipes from the guidebook. But I haven't decided yet. We'll see how things fill out. But this ingredient should work fine in shapeless recipes now.
So you guys should be good to go forward on this.

3TUSK commented:
If I read this right:
````java
public boolean isSimple() {
    return false;
}
`````
for all complex things like IngredientNBT and this one?

LexManos commented:
Yup, anything that isn't a simple item/subitem matching. Basically it's just a flag that tells it to use our matching system opposed to Vanilla's

3TUSK committed:
Merge remote-tracking branch 'remotes/upstream/1.12.x' into feature/fluid-ingredient

# Conflicts:
#	src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java

3TUSK committed:
Use `RecipeMatcher::findMatches`; add `isSimple` override

3TUSK committed:
`ICustomContainerCallback` - used for custom getContainerItem logic

This also effectively makes `FluidIngredient` a black-box.

3TUSK commented:
Updates:
* Merge conflicts resolved. 
* Adapt the `RecipeMatcher` thingy.
* @KnightMiner there is now an abstraction on custom container handling logic. Feedback wanted.
* These updates haven't been tested yet due to my dev env not working properly...

liach commented on src/main/java/net/minecraftforge/common/crafting/ICustomContainerCallback.java:
```diff
@@ -0,0 +1,45 @@
+     * @see net.minecraft.item.Item#getContainerItem(ItemStack)
+     * @see net.minecraftforge.common.ForgeHooks#getContainerItem(ItemStack)
+     */
+    ItemStack getContainer(ItemStack input);
```
`afterConsumption`

liach commented on src/main/java/net/minecraftforge/common/crafting/ICustomContainerCallback.java:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @see net.minecraftforge.common.ForgeHooks#getContainerItem(ItemStack)
+ */
+public interface ICustomContainerCallback
```
How about `IIngredientConsumption` which is more general (allows more more possibilities), such as consuming 1 durability for a hammer, etc.

liach commented on src/main/java/net/minecraftforge/oredict/ShapelessOreRecipe.java:
```diff
@@ -123,6 +125,41 @@ public boolean matches(@Nonnull InventoryCrafting inv, @Nonnull World world)
+        {
+            items.set(index, inv.getStackInSlot(index));
+        }
+        int[] mapping = RecipeMatcher.findMatches(items, this.input);
```
Call this based on whether this is simple.

LexManos replied:
Hes also not using it correctly.

3TUSK replied:
What did I missed?

LexManos replied:
https://github.com/MinecraftForge/MinecraftForge/commit/89db87dbfce4e4bd130f5daccac7cac495002275#diff-fb2aed325a1130a9796c31a2fee289b2R102
You need to only put in valid items.
You also nee to make sure the incoming arrays are equal in length to your inputs.

liach commented:
This one should work with the new matcher. @3TUSK Can you show a test screenshot?

3TUSK commented:
I am going to close this pull request due to real life stuff. @liach will take over this soon.
